### PR TITLE
Qualcomm AI Engine Direct - refactor quantizer

### PR DESCRIPTION
**Pull request by chunit-quic**
Summary:
- Refactor qnn quantizer
- Add 16bits default quantization setting

### CODE REVIEW

File: backends/qualcomm/qnn_quantizer.py [hunk 1]
```diff
@@ -226,847 +186,151 @@ def get_ptq_per_channel_weight_config() -> QuantizationConfig:
...
-                output_qspec=share_qparams_with_input_act0_qspec,
-                _annotated=True,
-            )
+        raise AssertionError(f"No quant config is implemented for op, {op}")
 
-    def _annotate_upsample2d(self, gm: torch.fx.GraphModule) -> None:
-        op_sources = [torch.nn.functional.interpolate, torch.nn.UpsamplingBilinear2d]
-        qconfig = self._get_quant_config(op_sources)
-        upsample_partitions = get_source_partitions(gm.graph, op_sources)
+    # TODO change to transform_for_annotation from the parent class when it's ready
```
Comment (cccclai) [inline backends/qualcomm/qnn_quantizer.py:309]: I think it's already there https://github.com/pytorch/pytorch/pull/113115

we can have a separate change to handle this todo though

File: backends/qualcomm/qnn_quantizer.py [hunk 2]
```diff
@@ -153,21 +104,17 @@ def get_default_qnn_ptq_config(
...
-    quantizer_configs = QnnQuantizerConfig(
-        enable_per_channel_conv_quant=enable_per_channel_conv_quant
-    )
-
-    return quantization_config, quantizer_configs
+    return quantization_config
 
 
-def get_16bit_qnn_ptq_config() -> Tuple[QuantizationConfig, QnnQuantizerConfig]:
+def get_default_16bit_qnn_ptq_config() -> QuantizationConfig:
```
Comment (cccclai) [inline backends/qualcomm/qnn_quantizer.py:110]: just curious, does qnn library mentioned in the recipe (https://pytorch.org/executorch/stable/build-run-qualcomm-ai-engine-direct-backend.html) supports 4 bit quantized linear?

**Review by cccclai — COMMENTED** [review]
Overall looks great! Maybe @jerryzh168 and @kimishpatel can take another look

**Comment by jerryzh168** [comment]
LGTM, how would you handle patterns like conv2d - relu though? will you put this under the conv2d key in OP_ANNOTATOR?

**Review by cccclai — APPROVED** [review]

File: backends/qualcomm/qnn_quantizer.py (hunk 1 above)
Comment (chunit-quic) [reply backends/qualcomm/qnn_quantizer.py:309]: Thanks for pointing out! we will prepare another PR for it.

File: backends/qualcomm/qnn_quantizer.py (hunk 2 above)
Comment (chunit-quic) [reply backends/qualcomm/qnn_quantizer.py:110]: 4 bit quantization is still WIP. We will introduce them to here once it finished.

**Comment by chunit-quic** [comment]
> LGTM, how would you handle patterns like conv2d - relu though? will you put this under the conv2d key in OP_ANNOTATOR?
Hi @jerryzh168,

Thank you for reviewing!
conv+relu pattern will be handled when [fold_relu](https://github.com/pytorch/executorch/blob/8305ce46f9ea1f9ef70b9214bf8056f23563cd5e/backends/qualcomm/utils/utils.py#L96) is ture. We don't need to handle this pattern specifically.

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Sending it back for

1. Change from histogram to minmax observer
2. understanding how you want to return supported ops

rest looks good

File: backends/qualcomm/qnn_quantizer.py (hunk 2 above)
Comment (cccclai) [reply backends/qualcomm/qnn_quantizer.py:110]: do we know an estimated timeline for this?

Comment (chunit-quic) [reply backends/qualcomm/qnn_quantizer.py:110]: Sorry that it might take a while to estimate the time. Because we are in the very early stage of investigation. we're still trying to understand things like QNN INT4 support coverage currently.

File: backends/qualcomm/qnn_quantizer.py [hunk 3]
```diff
@@ -116,23 +69,21 @@ def _derive_bias_qparams_fn(
...
     extra_args: Dict[str, Any] = {"eps": 2**-12}
 
     act_quantization_spec = QuantizationSpec(
         dtype=torch.uint8,
         quant_min=0,
-        quant_max=255,
+        quant_max=torch.iinfo(torch.uint8).max,
         qscheme=torch.per_tensor_affine,
-        observer_or_fake_quant_ctr=HistogramObserver.with_args(**extra_args),
+        observer_or_fake_quant_ctr=MovingAverageMinMaxObserver.with_args(**extra_args),
```
[the remark this answers is not in the document]
Comment (chunit-quic) [reply backends/qualcomm/qnn_quantizer.py:80]: Thank you for pointing out. We found that the accuracy is wrong when we use Histogram observer for mobilebert. Yet MovingAverageMean observer can obtain correct result, and it works on all the other test cases. Therefore we assume it might be fine to use MovingAverageMean as our default observer.
Would you recommend  to define a specific annotation configuration for mobilebert. If so we will make the config in  examples/qualcomm/scripts/ptq_mobilebert_fine_tune.py specifically

File: backends/qualcomm/qnn_quantizer.py [hunk 4]
```diff
@@ -226,847 +186,151 @@ def get_ptq_per_channel_weight_config() -> QuantizationConfig:
...
-                output_qspec=quantization_config.output_activation,
-                _annotated=True,
-            )
+        self.bit16_quant_config = quantization_config
 
-    def _annotate_linear(self, gm: torch.fx.GraphModule) -> None:  # noqa: C901
-        op_sources = [torch.nn.Linear, torch.nn.functional.linear]
-        linear_partitions = get_source_partitions(gm.graph, op_sources)
-        quantization_config = self._get_quant_config(op_sources)
+    def get_supported_ops(self) -> Set[OpOverload]:
```
[the remark this answers is not in the document]
Comment (chunit-quic) [reply backends/qualcomm/qnn_quantizer.py:263]: Right, we hope to make this quantizer have capability to handle 8 bits/16 bits, mixed precision quantization at the same time. Therefore we didn't categorize ops to a certain quantization config. There will be some following changes for the passes to support mixed precision

File: backends/qualcomm/qnn_quantizer.py [hunk 5]
```diff
@@ -226,847 +186,151 @@ def get_ptq_per_channel_weight_config() -> QuantizationConfig:
...
-            assert isinstance(input_node, Node)
-            input_qspec_map = {}
-            input_qspec_map[input_node] = qconfig.input_activation
-
-            upsample_node.meta[QUANT_ANNOTATION_KEY] = QuantizationAnnotation(
-                input_qspec_map=input_qspec_map,
-                output_qspec=qconfig.output_activation,
-                _annotated=True,
-            )
+            if (op := node.target) in self.bit8_quant_ops | self.bit16_quant_ops:
```
[the remark this answers is not in the document]
Comment (chunit-quic) [reply backends/qualcomm/qnn_quantizer.py]: Fixed. Thank you.

File: backends/qualcomm/qnn_quantizer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kimishpatel) [reply backends/qualcomm/qnn_quantizer.py:80]: If it is observed to work well for other models, then I dont have much concern. But if not i would modify specificaly for mobilebert

Also, would love to understand why histogram doesnt work for mobilbert but moving average works. This is interesting observation.

**Review by kimishpatel — APPROVED** [review]
